File: swift/Sources/AlloyAudio/Pack/PackSource.swift
import Foundation

/// Byte-origin seam for pack loading: manifest + per-zone encoded bytes.
/// Network stays at the host edge (injected), so PackLoader's logic stays
/// testable offline. Twin of web src/pack/pack-source.ts (canonical).
public protocol PackSource {
    func fetchManifest() async throws -> PackManifest
    func fetchZone(_ file: String) async throws -> Data
}

/// Minimal fetch surface (inject `URLSession` or a test double).
public typealias FetchFn = (String) async throws -> Data

/// Pack fetched from a base URL/path: `${base}/manifest.json`, `${base}/<file>`.
public struct BasePathPackSource: PackSource {
    private let base: String
    private let fetchFn: FetchFn

    public init(base: String, fetchFn: @escaping FetchFn) {
        self.base = base
        self.fetchFn = fetchFn
    }

    public func fetchManifest() async throws -> PackManifest {
        let data = try await fetchFn("\(base)/manifest.json")
        let manifest = try JSONDecoder().decode(PackManifest.self, from: data)
        let errors = validateManifest(manifest)
        if !errors.isEmpty {
            throw PackSourceError.invalidManifest(errors)
        }
        return manifest
    }

    public func fetchZone(_ file: String) async throws -> Data {
        try await fetchFn("\(base)/\(Self.encodeZonePath(file))")
    }

    /// Percent-encode each path segment of a manifest `file` entry.
    ///
    /// Zone filenames are content, not URLs, and real packs carry characters a
    /// URL parses as syntax — the piano pack's sharps are literally
    /// `D#4v12.m4a`, and an unencoded `#` is a FRAGMENT delimiter: the request
    /// silently becomes `<base>/D`, a static host answers with its index page,
    /// and the decoder then fails on markup. Encoding per segment (not the whole
    /// string) keeps `/` working for packs that nest zones in subdirectories.
    /// Twin: pack-source.ts (canonical).
    static func encodeZonePath(_ file: String) -> String {
        file.split(separator: "/", omittingEmptySubsequences: false)
            .map { segment in
                // .urlPathAllowed leaves `#` and `?` intact, which is exactly the
                // bug; subtract them so a segment can never introduce URL syntax.
                var allowed = CharacterSet.urlPathAllowed
                allowed.remove(charactersIn: "#?")
                return String(segment).addingPercentEncoding(withAllowedCharacters: allowed) ?? String(segment)
            }
            .joined(separator: "/")
    }
}

public enum PackSourceError: Error, Equatable {
    case invalidManifest([String])
}

File: web/packages/alloy-audio/src/pack/pack-source.ts
// Byte-origin + decode seams for pack loading. Both keep WebAudio/network at
// the host edge (injected), so the loader's logic stays testable offline.
// Twin: PackSource.swift + SampleDecoder.swift.

import { validateManifest, type PackManifest } from './manifest.js';

export type EncodedBytes = Uint8Array;

/** Mono PCM decoded from one .m4a. */
export interface DecodedPcm {
  sampleRate: number;
  data: Float32Array;
}

/** Decodes encoded (.m4a) bytes to mono PCM. Host-injected. */
export interface SampleDecoder {
  decode(bytes: EncodedBytes): Promise<DecodedPcm>;
}

/** Byte origin for a pack: manifest + per-zone encoded bytes. */
export interface PackSource {
  fetchManifest(): Promise<PackManifest>;
  fetchZone(file: string): Promise<EncodedBytes>;
}

/** Minimal fetch surface (inject globalThis.fetch or a test double). */
export type FetchFn = (url: string) => Promise<{
  json(): Promise<unknown>;
  arrayBuffer(): Promise<ArrayBuffer>;
}>;

/** Pack fetched from a base URL/path: `${base}/manifest.json`, `${base}/<file>`. */
export class BasePathPackSource implements PackSource {
  constructor(
    private readonly base: string,
    private readonly fetchFn: FetchFn,
  ) {}

  async fetchManifest(): Promise<PackManifest> {
    const res = await this.fetchFn(`${this.base}/manifest.json`);
    const manifest = (await res.json()) as PackManifest;
    const errors = validateManifest(manifest);
    if (errors.length > 0) throw new Error(`invalid manifest: ${errors.join('; ')}`);
    return manifest;
  }

  async fetchZone(file: string): Promise<EncodedBytes> {
    const res = await this.fetchFn(`${this.base}/${file}`);
    return new Uint8Array(await res.arrayBuffer());
  }
}

/** Minimal decode context (inject an AudioContext or a test double). */
export interface MinimalDecodeContext {
  decodeAudioData(data: ArrayBuffer): Promise<{
    sampleRate: number;
    numberOfChannels: number;
    getChannelData(channel: number): Float32Array;
  }>;
}

/** SampleDecoder backed by a WebAudio-like context; downmixes to mono. */
export class WebAudioDecoder implements SampleDecoder {
  constructor(private readonly ctx: MinimalDecodeContext) {}

  async decode(bytes: EncodedBytes): Promise<DecodedPcm> {
    const copy = bytes.slice();
    const buffer = await this.ctx.decodeAudioData(copy.buffer);
    const channels = buffer.numberOfChannels;
    const frames = buffer.getChannelData(0).length;
    const data = new Float32Array(frames);
    for (let c = 0; c < channels; c++) {
      const ch = buffer.getChannelData(c);
      for (let i = 0; i < frames; i++) data[i] += ch[i] / channels;
    }
    return { sampleRate: buffer.sampleRate, data };
  }
}
